In [1]:
import pandas as pd
import plotly.express as px
import numpy as np
from pprint import pprint

In [84]:
data1 = pd.read_csv("matches.csv") # ("./IPL Matches 2008-2020.csv")
data2 = pd.read_csv("deliveries.csv") # ("./IPL Ball-by-Ball 2008-2020.csv")
# data1.head()
print(len(data1), len(data2))

data1["team1"] = data1["team1"].replace("Rising Pune Supergiant", "Rising Pune Supergiants")
data1["team2"] = data1["team2"].replace("Rising Pune Supergiant", "Rising Pune Supergiants")
data1["winner"] = data1["winner"].replace("Rising Pune Supergiant", "Rising Pune Supergiants")

data1["team1"] = data1["team1"].replace("Delhi Daredevils", "Delhi Capitals")
data1["team2"] = data1["team2"].replace("Delhi Daredevils", "Delhi Capitals")
data1["winner"] = data1["winner"].replace("Delhi Daredevils", "Delhi Capitals")


756 179078


In [85]:
data1.columns
# data2[data2["is_super_over"] == 1].head(30)

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [86]:
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08      Rising Pune Supergiants   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1      Rising Pune Supergiants       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4               Delhi Capitals  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0      Rising Pune Supergiants            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [87]:
data1[data1["winner"] == np.nan]

Empty DataFrame
Columns: [id, season, city, date, team1, team2, toss_winner, toss_decision, result, dl_applied, winner, win_by_runs, win_by_wickets, player_of_match, venue, umpire1, umpire2, umpire3]
Index: []

In [88]:
teams = list(set(list(data1.team1)).union(set(list(data1.team2))))
teams

['Deccan Chargers',
 'Kochi Tuskers Kerala',
 'Rising Pune Supergiants',
 'Kolkata Knight Riders',
 'Mumbai Indians',
 'Royal Challengers Bangalore',
 'Rajasthan Royals',
 'Pune Warriors',
 'Chennai Super Kings',
 'Sunrisers Hyderabad',
 'Kings XI Punjab',
 'Gujarat Lions',
 'Delhi Capitals']

In [89]:
data1.winner.unique()

array(['Sunrisers Hyderabad', 'Rising Pune Supergiants',
       'Kolkata Knight Riders', 'Kings XI Punjab',
       'Royal Challengers Bangalore', 'Mumbai Indians', 'Delhi Capitals',
       'Gujarat Lions', 'Chennai Super Kings', 'Rajasthan Royals',
       'Deccan Chargers', 'Pune Warriors', 'Kochi Tuskers Kerala', nan],
      dtype=object)

# Objective 1a: Deciding the best teams in IPL history.

In this regard, we can model a reasonable assumption: "The teams with more wins are more likely to be the best teams, as they are more likely to perform better." 

In [90]:
team_wins = {}
team_losses = {}
team_ties = {}
# data1 = data1[data1["winner"] != np.nan]

nm1 = np.nan # Ties
nm2 = np.nan # Losses
nm = np.nan # Wins

for i in range(len(data1)):
    nm = data1.iloc[i,:].winner
    
    if data1.iloc[i,:].team2 == data1.iloc[i,:].winner:
        nm1 = data1.iloc[i,:].team1
    elif data1.iloc[i,:].team1 == data1.iloc[i,:].winner:
        nm1 = data1.iloc[i,:].team2

    if nm != np.nan:
        if nm in team_wins.keys(): 
            team_wins[nm] += 1
        else:
            team_wins[nm] = 1
    
    if nm1 != np.nan:
        if nm1 in team_losses.keys(): 
            team_losses[nm1] += 1
        else:
            team_losses[nm1] = 1
    
    if nm == np.nan:
        t1 = data1.iloc[i,:].team1
        t2 = data1.iloc[i,:].team2

        if t1 in team_ties.keys():
            team_ties[t1] +=1
        else:
            team_ties[t1] = 1

        if t2 in team_ties.keys():
            team_ties[t2] +=1
        else:
            team_ties[t2] = 1
    
del team_wins[np.nan]
# del team_losses[np.nan]
print("Win counts:\n")
pprint(sorted(team_wins.items(), key= lambda x: x[1], reverse=True))
print("----------------------------------\n")
print("Loss counts:\n")
pprint(sorted(team_losses.items(), key= lambda x: x[1], reverse=True))
print("----------------------------------\n")
print("Tie counts:\n")
pprint(sorted(team_ties.items(), key= lambda x: x[1], reverse=True))
print("----------------------------------\n")

Win counts:

[('Mumbai Indians', 109),
 ('Chennai Super Kings', 100),
 ('Kolkata Knight Riders', 92),
 ('Royal Challengers Bangalore', 84),
 ('Kings XI Punjab', 82),
 ('Delhi Capitals', 77),
 ('Rajasthan Royals', 75),
 ('Sunrisers Hyderabad', 58),
 ('Deccan Chargers', 29),
 ('Rising Pune Supergiants', 15),
 ('Gujarat Lions', 13),
 ('Pune Warriors', 12),
 ('Kochi Tuskers Kerala', 6)]
----------------------------------

Loss counts:

[('Delhi Capitals', 99),
 ('Kings XI Punjab', 96),
 ('Royal Challengers Bangalore', 93),
 ('Kolkata Knight Riders', 87),
 ('Mumbai Indians', 78),
 ('Rajasthan Royals', 70),
 ('Chennai Super Kings', 64),
 ('Sunrisers Hyderabad', 50),
 ('Deccan Chargers', 46),
 ('Pune Warriors', 33),
 ('Gujarat Lions', 17),
 ('Rising Pune Supergiants', 15),
 ('Kochi Tuskers Kerala', 8)]
----------------------------------

Tie counts:

[]
----------------------------------



In [91]:
team_stats = pd.DataFrame({"team": [], "wins": [], "losses": [], "ties": []})

for team in team_wins.keys():
    team_stats = team_stats.append({"team": team, "wins": team_wins[team], "losses": team_losses[team]}, ignore_index = True)

team_stats.fillna(0)

team   wins  losses  ties
0           Sunrisers Hyderabad   58.0    50.0   0.0
1       Rising Pune Supergiants   15.0    15.0   0.0
2         Kolkata Knight Riders   92.0    87.0   0.0
3               Kings XI Punjab   82.0    96.0   0.0
4   Royal Challengers Bangalore   84.0    93.0   0.0
5                Mumbai Indians  109.0    78.0   0.0
6                Delhi Capitals   77.0    99.0   0.0
7                 Gujarat Lions   13.0    17.0   0.0
8           Chennai Super Kings  100.0    64.0   0.0
9              Rajasthan Royals   75.0    70.0   0.0
10              Deccan Chargers   29.0    46.0   0.0
11                Pune Warriors   12.0    33.0   0.0
12         Kochi Tuskers Kerala    6.0     8.0   0.0

In [92]:
team_stats_disp = team_stats[["team", "wins", "losses"]]
team_stats_disp = team_stats_disp.sort_values("wins", ascending=False)
team_stats_disp

team   wins  losses
5                Mumbai Indians  109.0    78.0
8           Chennai Super Kings  100.0    64.0
2         Kolkata Knight Riders   92.0    87.0
4   Royal Challengers Bangalore   84.0    93.0
3               Kings XI Punjab   82.0    96.0
6                Delhi Capitals   77.0    99.0
9              Rajasthan Royals   75.0    70.0
0           Sunrisers Hyderabad   58.0    50.0
10              Deccan Chargers   29.0    46.0
1       Rising Pune Supergiants   15.0    15.0
7                 Gujarat Lions   13.0    17.0
11                Pune Warriors   12.0    33.0
12         Kochi Tuskers Kerala    6.0     8.0

In [93]:
px.bar(team_stats_disp, x="team", y=["wins", "losses"], title="Team Stats, ranked by success in IPL (2008-2019)", labels={
    "team": "Team Name",
    "value": "Match Results"
}).show()

# Inference 1a:

From the above graph, we can infer that the top 3 successful teams in IPL are:
- Mumbai Indians (109 wins, 78 losses)
- Chennai Super Kings(100 wins, 64 losses)
- Kolkata Knight Riders(92 wins, 87 losses)

# Objective 1b: To find the top-performing players in IPL history.



## Finding the top-performing batsmen

In [94]:
batsmen = sorted(list(set(list(data2["batsman"]))))

balls_faced_index = {}

for i in batsmen:
    balls_faced_index[i] = len(data2[data2["batsman"] == i])

runs_index = {}

for i in batsmen:
    runs_index[i] = data2[data2["batsman"] == i].batsman_runs.sum()

# pprint(runs_index)
len(runs_index)

top_batsmen = sorted(runs_index.items(), key = lambda x: x[1], reverse=True)
# pprint(top_batsmen[:20])

strike_rate_index = {}

for i in runs_index.keys():
    strike_rate_index[i] = runs_index[i]/balls_faced_index[i] * 100

# strike_rate_index



In [95]:
batsmen_matches_index = {}
buf = None
for i in batsmen:
    batsmen_matches_index[i] = len(set(list(data2[data2["batsman"] == i].match_id)))

In [96]:
dict(sorted(batsmen_matches_index.items(), key=lambda x: x[1], reverse=True)[:20])

{'SK Raina': 189,
 'RG Sharma': 182,
 'MS Dhoni': 170,
 'RV Uthappa': 170,
 'V Kohli': 169,
 'KD Karthik': 162,
 'S Dhawan': 158,
 'YK Pathan': 153,
 'G Gambhir': 151,
 'AB de Villiers': 142,
 'AT Rayudu': 138,
 'PA Patel': 136,
 'AM Rahane': 132,
 'KA Pollard': 132,
 'SR Watson': 130,
 'DA Warner': 126,
 'Yuvraj Singh': 126,
 'RA Jadeja': 125,
 'CH Gayle': 124,
 'MK Pandey': 119}

In [97]:
batsmen_df = pd.DataFrame({"Name": [], "Matches": [], "Runs": [], "Balls faced": [], "Strike Rate": []})

for i in batsmen:
    batsmen_df = batsmen_df.append({"Name": i, "Matches": batsmen_matches_index[i], "Runs": runs_index[i], "Balls faced": balls_faced_index[i], "Strike Rate": strike_rate_index[i]}, ignore_index=True)

batsmen_df = batsmen_df.sort_values("Runs", ascending=False)

In [98]:
batsmen_df.iloc[:20]

Name  Matches    Runs  Balls faced  Strike Rate
486         V Kohli    169.0  5434.0       4211.0   129.042983
428        SK Raina    189.0  5415.0       4044.0   133.902077
367       RG Sharma    182.0  4914.0       3816.0   128.773585
112       DA Warner    126.0  4741.0       3398.0   139.523249
392        S Dhawan    158.0  4632.0       3776.0   122.669492
92         CH Gayle    124.0  4560.0       3131.0   145.640370
290        MS Dhoni    170.0  4477.0       3318.0   134.930681
384      RV Uthappa    170.0  4446.0       3492.0   127.319588
26   AB de Villiers    142.0  4428.0       2977.0   148.740343
147       G Gambhir    151.0  4223.0       3524.0   119.835414
42        AM Rahane    132.0  3850.0       3215.0   119.751166
218      KD Karthik    162.0  3688.0       2890.0   127.612457
444       SR Watson    130.0  3614.0       2639.0   136.945813
53        AT Rayudu    138.0  3326.0       2681.0   124.058187
509       YK Pathan    153.0  3241.0       2334.0   138.860326
72      BB McCullum    109.0  2893.0       2272.0   127.332746
329        PA Patel    136.0  2874.0       2444.0   117.594108
280       MK Pandey    119.0  2872.0       2425.0   118.432990
213      KA Pollard    132.0  2784.0       1964.0   141.751527
514    Yuvraj Singh    126.0  2765.0       2207.0   125.283190

## Find the top performing bowlers

In [99]:
bowlers = sorted(list(set(list(data2["bowler"]))))

deliveries_index = {}
for i in bowlers:
    deliveries_index[i] = len(data2[data2["bowler"] == i])


In [100]:
# Verification if total sum of all deliveries match number of rows of deliveries.csv

sumd = 0 
for i in bowlers:
    sumd += deliveries_index[i]

sumd

179078

In [101]:
# pprint(deliveries_index)
pprint(sorted(deliveries_index.items(), key = lambda x: x[1], reverse=True))

[('Harbhajan Singh', 3451),
 ('A Mishra', 3172),
 ('PP Chawla', 3157),
 ('R Ashwin', 3016),
 ('SL Malinga', 2974),
 ('DJ Bravo', 2711),
 ('B Kumar', 2707),
 ('P Kumar', 2637),
 ('UT Yadav', 2605),
 ('SP Narine', 2600),
 ('RA Jadeja', 2541),
 ('Z Khan', 2276),
 ('DW Steyn', 2207),
 ('R Vinay Kumar', 2186),
 ('SR Watson', 2137),
 ('IK Pathan', 2113),
 ('I Sharma', 1999),
 ('A Nehra', 1974),
 ('PP Ojha', 1945),
 ('RP Singh', 1874),
 ('YS Chahal', 1841),
 ('DS Kulkarni', 1830),
 ('JA Morkel', 1807),
 ('JH Kallis', 1799),
 ('JJ Bumrah', 1798),
 ('MM Sharma', 1770),
 ('AR Patel', 1745),
 ('M Morkel', 1699),
 ('R Bhatia', 1661),
 ('AB Dinda', 1589),
 ('M Muralitharan', 1581),
 ('L Balaji', 1574),
 ('SK Trivedi', 1557),
 ('Sandeep Sharma', 1546),
 ('JD Unadkat', 1541),
 ('MM Patel', 1382),
 ('Shakib Al Hasan', 1359),
 ('MJ McClenaghan', 1346),
 ('CH Morris', 1335),
 ('MG Johnson', 1301),
 ('JP Faulkner', 1287),
 ('S Nadeem', 1283),
 ('KA Pollard', 1276),
 ('Imran Tahir', 1249),
 ('SK Warne', 1

In [102]:
print(data2["dismissal_kind"].unique())
kinds = ["caught", "bowled", "lbw", "caught and bowled", "stumped"]

[nan 'caught' 'bowled' 'run out' 'lbw' 'caught and bowled' 'stumped'
 'retired hurt' 'hit wicket' 'obstructing the field']


In [103]:
wickets_index = {}

for i in bowlers:
    wickets_index[i] = len(data2.loc[(data2["player_dismissed"].isin(batsmen)) & (data2["bowler"] == i) & (data2["dismissal_kind"].isin(kinds))])



In [104]:
pprint(sorted(wickets_index.items(), key = lambda x: x[1], reverse=True))

[('SL Malinga', 170),
 ('A Mishra', 156),
 ('Harbhajan Singh', 149),
 ('PP Chawla', 149),
 ('DJ Bravo', 147),
 ('B Kumar', 133),
 ('R Ashwin', 125),
 ('SP Narine', 122),
 ('UT Yadav', 119),
 ('RA Jadeja', 108),
 ('A Nehra', 106),
 ('R Vinay Kumar', 105),
 ('Z Khan', 102),
 ('YS Chahal', 100),
 ('DW Steyn', 95),
 ('SR Watson', 92),
 ('MM Sharma', 91),
 ('P Kumar', 90),
 ('RP Singh', 90),
 ('PP Ojha', 89),
 ('DS Kulkarni', 86),
 ('JA Morkel', 84),
 ('Sandeep Sharma', 83),
 ('JJ Bumrah', 82),
 ('IK Pathan', 80),
 ('Imran Tahir', 79),
 ('JD Unadkat', 77),
 ('M Morkel', 77),
 ('L Balaji', 76),
 ('MM Patel', 74),
 ('I Sharma', 71),
 ('R Bhatia', 71),
 ('AR Patel', 70),
 ('CH Morris', 70),
 ('MJ McClenaghan', 70),
 ('AB Dinda', 69),
 ('JH Kallis', 65),
 ('SK Trivedi', 65),
 ('M Muralitharan', 64),
 ('MG Johnson', 62),
 ('JP Faulkner', 61),
 ('Shakib Al Hasan', 59),
 ('SK Warne', 57),
 ('KA Pollard', 56),
 ('Rashid Khan', 55),
 ('AD Russell', 54),
 ('KV Sharma', 54),
 ('S Kaul', 49),
 ('SB Jak

In [105]:
runs_conceded_index = {}

for i in bowlers:
    runs_conceded_index[i] = data2.loc[(data2["bowler"]) == i].total_runs.sum()

In [106]:
dict(sorted(runs_conceded_index.items(), key = lambda x: x[1], reverse=True)[:20])

{'PP Chawla': 4153,
 'Harbhajan Singh': 4050,
 'A Mishra': 3850,
 'DJ Bravo': 3733,
 'UT Yadav': 3640,
 'SL Malinga': 3511,
 'R Ashwin': 3391,
 'P Kumar': 3342,
 'B Kumar': 3264,
 'RA Jadeja': 3221,
 'R Vinay Kumar': 3043,
 'SP Narine': 2939,
 'Z Khan': 2860,
 'SR Watson': 2751,
 'IK Pathan': 2711,
 'I Sharma': 2682,
 'A Nehra': 2537,
 'DS Kulkarni': 2461,
 'MM Sharma': 2457,
 'DW Steyn': 2454}

In [107]:
average_index = {}

for i in bowlers:
    average_index[i] = runs_conceded_index[i]/wickets_index[i]

<ipython-input-107-be16ba59c22c>:4: RuntimeWarning:

divide by zero encountered in long_scalars



In [108]:
average_index["SL Malinga"]

20.652941176470588

In [109]:
bowlers_matches_index = {}
for i in bowlers:
    bowlers_matches_index[i] = len(set(list(data2[data2["bowler"] == i].match_id)))

In [110]:
bowlers_df = pd.DataFrame({"Name": [], "Matches": [], "Wickets": [], "Runs conceded": [], "Average": []})

for i in bowlers:
    bowlers_df = bowlers_df.append({"Name": i, "Matches": bowlers_matches_index[i], "Wickets": wickets_index[i], "Runs conceded": runs_conceded_index[i], "Average": average_index[i]}, ignore_index=True)

bowlers_df = bowlers_df.sort_values("Wickets", ascending=False)

In [111]:
bowlers_df

Name  Matches  Wickets  Runs conceded    Average
340       SL Malinga    122.0    170.0         3511.0  20.652941
7           A Mishra    147.0    156.0         3850.0  24.679487
126  Harbhajan Singh    157.0    149.0         4050.0  27.181208
264        PP Chawla    156.0    149.0         4153.0  27.872483
93          DJ Bravo    131.0    147.0         3733.0  25.394558
..               ...      ...      ...            ...        ...
366      Sunny Gupta      1.0      0.0           47.0        inf
139           J Dala      1.0      0.0           35.0        inf
51          B Chipli      1.0      0.0           20.0        inf
306           S Dube      2.0      0.0            9.0        inf
243    NJ Rimmington      1.0      0.0           20.0        inf

[405 rows x 5 columns]

In [112]:
bowlers_df.loc[bowlers_df["Matches"] >= 50].sort_values("Average", ascending=True)[:10]

Name  Matches  Wickets  Runs conceded    Average
340      SL Malinga    122.0    170.0         3511.0  20.652941
134     Imran Tahir     55.0     79.0         1656.0  20.962025
225        MM Patel     63.0     74.0         1733.0  23.418919
358  Sandeep Sharma     68.0     83.0         1967.0  23.698795
9           A Nehra     88.0    106.0         2537.0  23.933962
347       SP Narine    109.0    122.0         2939.0  24.090164
402       YS Chahal     83.0    100.0         2416.0  24.160000
53          B Kumar    117.0    133.0         3264.0  24.541353
7          A Mishra    147.0    156.0         3850.0  24.679487
93         DJ Bravo    131.0    147.0         3733.0  25.394558

In [113]:
pprint(bowlers_df[:10].Name)

340         SL Malinga
7             A Mishra
126    Harbhajan Singh
264          PP Chawla
93            DJ Bravo
53             B Kumar
269           R Ashwin
347          SP Narine
381           UT Yadav
282          RA Jadeja
Name: Name, dtype: object


# Inference 1b: 

The top 10 batsmen in IPL history (going by runs) are:

- V Kohli
- SK Raina
- RG Sharma
- DA Warner
- S Dhawan
- CH Gayle
- MS Dhoni
- RV Uthappa
- AB de Villiers
- G Gambhir

The top 10 batsmen in IPL history (going by wickets) are:

- SL Malinga
- A Mishra
- Harbhajan Singh
- PP Chawla
- DJ Bravo
- B Kumar
- R Ashwin
- SP Narine
- UT Yadav
- RA Jadeja

# Objective 2: To find the most favourable players to be ambassadors of different companies for promotions

The top-performing player in each match is regarded as the Man of the Match (MoM). And popular So an argument can be modelled optimally: "The players with more MoM rewards are more likely to be popular."

In [32]:
MoM_names = list(data1.player_of_match)

MoM_index = {}

for i in MoM_names:
    if i in MoM_index.keys():
        MoM_index[i] += 1
    else:
        MoM_index[i] = 1
MoM_index

# print(len(MoM_index.keys()))
top_MoM = pd.DataFrame({"Name": [], "MoM awards": []})

for i in MoM_index.keys():
    top_MoM = top_MoM.append({"Name": i, "MoM awards": MoM_index[i]}, ignore_index=True)

top_MoM = top_MoM.sort_values("MoM awards", ascending=False)

top_MoM.iloc[:20, :]

Name  MoM awards
18         CH Gayle        21.0
90   AB de Villiers        20.0
30        RG Sharma        17.0
23         MS Dhoni        17.0
29        DA Warner        17.0
55        YK Pathan        16.0
52        SR Watson        15.0
21         SK Raina        14.0
25        G Gambhir        13.0
48       MEK Hussey        12.0
150       AM Rahane        12.0
138         V Kohli        12.0
53         V Sehwag        11.0
36         DR Smith        11.0
184      AD Russell        11.0
74         A Mishra        11.0
98        JH Kallis        10.0
11       KA Pollard        10.0
62         SE Marsh         9.0
43        AT Rayudu         9.0

# Head-to-Head (H2H) matchups

In [33]:
h2h = data1[["team1", "team2", "winner", "venue"]]
h2h_data = pd.DataFrame({"team1": [], "team2": [], "matchups": [], "wins for team 1": [], "wins for team 2": [], "tie(s)": [], "probability of team1 winning": [], "probability of team2 winning": [], "probability of tie": [], "venue": []})

h2h.reset_index(drop=True, inplace=True)
buf = None
explored = []
for i in teams:
    ct1 = 0
    ct2 = 0
    ct3 = 0
    sct = 0
    for j in teams:
        if i != j:
            # print(i, j)
            buf = h2h.loc[(h2h["team1"] == i) & (h2h["team2"] == j)]
            for k in range(len(buf)):
                # print(buf.iloc[k].winner)
                if buf.iloc[k].winner == i:
                    ct1 += 1
                elif buf.iloc[k].winner == j:
                    ct2 += 1
                else:
                    ct3 += 1
            sct = ct1+ct2+ct3
            if sct == 0: sct = 1
            h2h_data = h2h_data.append({"team1": i, "team2": j, "matchups": ct1+ct2+ct3, "wins for team 1": ct1, "wins for team 2": ct2, "tie(s)": ct3, "probability of team1 winning": ct1/sct, "probability of team2 winning": ct2/sct, "probability of tie": ct3/sct, "venue": list(buf.venue)}, ignore_index=True)
        ct1 = 0
        ct2 = 0
        ct3 = 0

h2h_data

team1                        team2  matchups  wins for team 1  \
0    Deccan Chargers         Kochi Tuskers Kerala       1.0              1.0   
1    Deccan Chargers      Rising Pune Supergiants       0.0              0.0   
2    Deccan Chargers        Kolkata Knight Riders       2.0              0.0   
3    Deccan Chargers               Mumbai Indians       5.0              3.0   
4    Deccan Chargers  Royal Challengers Bangalore       8.0              5.0   
..               ...                          ...       ...              ...   
151   Delhi Capitals                Pune Warriors       2.0              1.0   
152   Delhi Capitals          Chennai Super Kings       8.0              2.0   
153   Delhi Capitals          Sunrisers Hyderabad       8.0              2.0   
154   Delhi Capitals              Kings XI Punjab      11.0              2.0   
155   Delhi Capitals                Gujarat Lions       0.0              0.0   

     wins for team 2  tie(s)  probability of team1 winning  \
0                0.0     0.0                      1.000000   
1                0.0     0.0                      0.000000   
2                2.0     0.0                      0.000000   
3                2.0     0.0                      0.600000   
4                3.0     0.0                      0.625000   
..               ...     ...                           ...   
151              0.0     1.0                      0.500000   
152              6.0     0.0                      0.250000   
153              6.0     0.0                      0.250000   
154              9.0     0.0                      0.181818   
155              0.0     0.0                      0.000000   

     probability of team2 winning  probability of tie  \
0                        0.000000                 0.0   
1                        0.000000                 0.0   
2                        1.000000                 0.0   
3                        0.400000                 0.0   
4                        0.375000                 0.0   
..                            ...                 ...   
151                      0.000000                 0.5   
152                      0.750000                 0.0   
153                      0.750000                 0.0   
154                      0.818182                 0.0   
155                      0.000000                 0.0   

                                                 venue  
0                                      [Nehru Stadium]  
1                                                   []  
2                     [Eden Gardens, Barabati Stadium]  
3    [Kingsmead, SuperSport Park, Wankhede Stadium,...  
4    [Rajiv Gandhi International Stadium, Uppal, Ne...  
..                                                 ...  
151  [Feroz Shah Kotla, Shaheed Veer Narayan Singh ...  
152  [Feroz Shah Kotla, Kingsmead, Feroz Shah Kotla...  
153  [Feroz Shah Kotla, Rajiv Gandhi International ...  
154  [Feroz Shah Kotla, Punjab Cricket Association ...  
155                                                 []  

[156 rows x 10 columns]

### Verifying that the data filtering is correct

In [34]:
h2h_data["wins for team 2"].sum() + h2h_data["wins for team 1"].sum() + h2h_data["tie(s)"].sum() # should be same as total matches played (756)

756.0

### Finding tied matches:

In [35]:
h2h_data[h2h_data["tie(s)"] > 0.0]

team1                        team2  matchups  \
65   Royal Challengers Bangalore             Rajasthan Royals      11.0   
149               Delhi Capitals  Royal Challengers Bangalore      13.0   
151               Delhi Capitals                Pune Warriors       2.0   

     wins for team 1  wins for team 2  tie(s)  probability of team1 winning  \
65               3.0              6.0     2.0                      0.272727   
149              4.0              8.0     1.0                      0.307692   
151              1.0              0.0     1.0                      0.500000   

     probability of team2 winning  probability of tie  \
65                       0.545455            0.181818   
149                      0.615385            0.076923   
151                      0.000000            0.500000   

                                                 venue  
65   [M Chinnaswamy Stadium, Newlands, SuperSport P...  
149  [Feroz Shah Kotla, New Wanderers Stadium, M Ch...  
151  [Feroz Shah Kotla, Shaheed Veer Narayan Singh ...

In [164]:
# data1[(data1["team1"] == "Kings XI Punjab") & (data1["team2"] == "Kolkata Knight Riders")] # Verifying correctness of findings

In [36]:
h2h_data.sort_values("matchups", ascending=False)

team1                        team2  matchups  \
75          Rajasthan Royals        Kolkata Knight Riders      17.0   
59            Mumbai Indians               Delhi Capitals      17.0   
123          Kings XI Punjab        Kolkata Knight Riders      17.0   
55            Mumbai Indians          Chennai Super Kings      16.0   
40     Kolkata Knight Riders  Royal Challengers Bangalore      16.0   
..                       ...                          ...       ...   
37     Kolkata Knight Riders         Kochi Tuskers Kerala       0.0   
35   Rising Pune Supergiants               Delhi Capitals       0.0   
108      Sunrisers Hyderabad              Deccan Chargers       0.0   
133            Gujarat Lions         Kochi Tuskers Kerala       0.0   
155           Delhi Capitals                Gujarat Lions       0.0   

     wins for team 1  wins for team 2  tie(s)  probability of team1 winning  \
75               8.0              9.0     0.0                      0.470588   
59              10.0              7.0     0.0                      0.588235   
123              5.0             12.0     0.0                      0.294118   
55              10.0              6.0     0.0                      0.625000   
40               7.0              9.0     0.0                      0.437500   
..               ...              ...     ...                           ...   
37               0.0              0.0     0.0                      0.000000   
35               0.0              0.0     0.0                      0.000000   
108              0.0              0.0     0.0                      0.000000   
133              0.0              0.0     0.0                      0.000000   
155              0.0              0.0     0.0                      0.000000   

     probability of team2 winning  probability of tie  \
75                       0.529412                 0.0   
59                       0.411765                 0.0   
123                      0.705882                 0.0   
55                       0.375000                 0.0   
40                       0.562500                 0.0   
..                            ...                 ...   
37                       0.000000                 0.0   
35                       0.000000                 0.0   
108                      0.000000                 0.0   
133                      0.000000                 0.0   
155                      0.000000                 0.0   

                                                 venue  
75   [Sawai Mansingh Stadium, Newlands, Kingsmead, ...  
59   [Wankhede Stadium, Feroz Shah Kotla, Dr DY Pat...  
123  [Eden Gardens, Punjab Cricket Association IS B...  
55   [Newlands, St George's Park, Wankhede Stadium,...  
40   [Eden Gardens, M Chinnaswamy Stadium, Eden Gar...  
..                                                 ...  
37                                                  []  
35                                                  []  
108                                                 []  
133                                                 []  
155                                                 []  

[156 rows x 10 columns]

## Performance on home ground

In [45]:
fig1a = px.bar(h2h_data, x = "team1", y="wins for team 1", text="wins for team 1", color="team2", labels={
    "team1": "Team",
    "wins for team 1": "Wins",
    "team2": "Opponent",

}, title="Stacked head-to-head wins at home", width=1000, height=800)
fig1a.update_layout(uniformtext_minsize=8)
fig1a.show()

In [49]:
fig2a = px.bar(h2h_data, x = "team1" , y="wins for team 2", text="wins for team 2", color="team2", labels={
    "team1": "Team",
    "wins for team 2": "Losses",
    "team2": "Opponent",

}, title="Stacked head-to-head losses at home", width=1000, height=800)
fig2a.update_layout(uniformtext_minsize=8)
fig2a.show()

In [50]:
figA = px.bar(h2h_data, x = "team1", y=["wins for team 1", "wins for team 2"], color="team2", labels={
    "team1": "Team",
    "value": "Stats (wins | losses)",
    "team2": "Opponent",

}, title="Stacked head-to-head performance at home", width=1000, height=800)

figA.show()

## Performance in away matches

In [53]:
fig1b = px.bar(h2h_data, x = "team2", y="wins for team 2", text="wins for team 2", color="team1", labels={
    "team2": "Team",
    "wins for team 2": "Wins",
    "team1": "Opponent",

}, title="Stacked head-to-head wins at away matches", width=1000, height=800)
fig1b.update_layout(uniformtext_minsize=8)
fig1b.show()

In [54]:
fig2b = px.bar(h2h_data, x = "team2", y="wins for team 1", text="wins for team 1", color="team1", labels={
    "team2": "Team",
    "wins for team 1": "Losses",
    "team1": "Opponent",

}, title="Stacked head-to-head losses at away matches", width=1000, height=800)
fig2b.update_layout(uniformtext_minsize=8)
fig2b.show()

In [55]:
figB = px.bar(h2h_data, x = "team2", y=["wins for team 2", "wins for team 1"], color="team1", labels={
    "team2": "Team",
    "value": "Stats (wins | losses)",
    "team1": "Opponent",

}, title="Stacked head-to-head performance at away matches", width=1000, height=800)

figB.show()

In [40]:
data1.loc[(data1["team1"] == "Gujarat Lions") & (data1["winner"] == "Gujarat Lions")]

id  season   city        date          team1           team2  \
598  599    2016  Delhi  2016-04-27  Gujarat Lions  Delhi Capitals   

          toss_winner toss_decision  result  dl_applied         winner  \
598  Delhi Daredevils         field  normal           0  Gujarat Lions   

     win_by_runs  win_by_wickets player_of_match             venue    umpire1  \
598            1               0       CH Morris  Feroz Shah Kotla  M Erasmus   

    umpire2 umpire3  
598  S Ravi     NaN

In [58]:
len(data1.loc[(data1["team1"] == "Gujarat Lions")])

14

In [73]:
data3 = pd.read_csv("./IPL Matches 2008-2020.csv")
data4 = pd.read_csv("./IPL Ball-by-Ball 2008-2020.csv")

c = 0
cl = []
for i in data3.columns:
    if i in data1.columns:
        c += 1
        cl.append(i)
        
print(c)
print(cl)
# print(data1.columns)
# print(data3.columns)

13
['id', 'city', 'date', 'player_of_match', 'venue', 'team1', 'team2', 'toss_winner', 'toss_decision', 'winner', 'result', 'umpire1', 'umpire2']
